core/runtime/common/runtime_instances_pool.cpp

```diff
@@ -74,63 +74,65 @@ namespace kagome::runtime {
   RuntimeInstancesPool::instantiateFromCode(const CodeHash &code_hash,
                                             common::BufferView code_zstd) {
+    std::optional<std::reference_wrapper<InstancePool>> pool_opt{};
+
     std::unique_lock lock{pools_mtx_};
-    auto pool_opt = pools_.get(code_hash);
+    pool_opt = pools_.get(code_hash);
 
     if (!pool_opt) {
       lock.unlock();
-      if (auto future = getFutureCompiledModule(code_hash)) {
-        lock.lock();
-        pool_opt = pools_.get(code_hash);
+      auto module_res = tryCompileModule(code_hash, code_zstd);
+      lock.lock();
+      if (module_res) {
+        if (auto pool = pools_.get(code_hash)) {
+          pool_opt = std::ref(pool->get());
+        } else {
+          pool_opt = std::ref(
+              pools_.put(code_hash, InstancePool{module_res.value(), {}}));
+        }
       } else {
-        OUTCOME_TRY(module, tryCompileModule(code_hash, code_zstd));
-        BOOST_ASSERT(module != nullptr);
-        lock.lock();
-        pool_opt = std::ref(pools_.put(code_hash, InstancePool{module, {}}));
+        return module_res.error();
       }
     }
+    BOOST_ASSERT(pool_opt);
     auto instance = pool_opt->get().instantiate(lock);
     return std::make_shared<BorrowedInstance>(
         weak_from_this(), code_hash, std::move(instance));
   }
 
-  std::optional<std::shared_future<RuntimeInstancesPool::CompilationResult>>
-  RuntimeInstancesPool::getFutureCompiledModule(
-      const CodeHash &code_hash) const {
-    std::unique_lock l{compiling_modules_mtx_};
-    auto iter = compiling_modules_.find(code_hash);
-    if (iter == compiling_modules_.end()) {
-      return std::nullopt;
-    }
-    auto future = iter->second;
-    l.unlock();
-    return future;
-  }
-
   RuntimeInstancesPool::CompilationResult
   RuntimeInstancesPool::tryCompileModule(const CodeHash &code_hash,
                                          common::BufferView code_zstd) {
     std::unique_lock l{compiling_modules_mtx_};
+    if (auto iter = compiling_modules_.find(code_hash);
+        iter != compiling_modules_.end()) {
+      std::shared_future<CompilationResult> future = iter->second;
+      l.unlock();
+      return future.get();
+    }
     std::promise<CompilationResult> promise;
-    auto [iter, inserted] =
+    auto [iter, is_inserted] =
         compiling_modules_.insert({code_hash, promise.get_future()});
-    BOOST_ASSERT(inserted);
+    BOOST_ASSERT(is_inserted);
     l.unlock();
 
     common::Buffer code;
-    CompilationResult res{nullptr};
+    std::optional<CompilationResult> res;
     if (!uncompressCodeIfNeeded(code_zstd, code)) {
       res = CompilationError{"Failed to uncompress code"};
     } else {
       res = common::map_result(module_factory_->make(code), [](auto &&module) {
         return std::shared_ptr<const Module>(module);
       });
     }
-    promise.set_value(res);
+    BOOST_ASSERT(res);
 
-    l.lock();
+    std::unique_lock ll{compiling_modules_mtx_};
```

Comment:
```suggestion
    l.lock();
```

Reply:
Eh, I dislike these manual locks-unlocks, but it will probably be more consistent, okay.